public/js/utils/editor.js

```diff
@@ -22,7 +26,154 @@ function shouldShowFooter(selectedSource) {
   return shouldShowPrettyPrint(selectedSource);
 }
 
+function setText(editor, text) {
+  if (!text || !editor) {
+    return;
+  }
+
+  editor.setText(text);
+}
+
+function setMode(editor, sourceText) {
+  const contentType = sourceText.get("contentType");
+
+  if (contentType.includes("javascript")) {
+    editor.setMode({ name: "javascript" });
+  } else if (contentType === "text/wasm") {
+    editor.setMode({ name: "text" });
+  } else if (sourceText.get("text").match(/^\s*</)) {
+    // Use HTML mode for files in which the first non whitespace
+    // character is `<` regardless of extension.
+    editor.setMode({ name: "htmlmixed" });
+  } else {
+    editor.setMode({ name: "text" });
+  }
+}
+
+function showMessage(editor, msg) {
+  editor.replaceDocument(editor.createDocument());
+  setText(editor, msg);
+  editor.setMode({ name: "text" });
+}
+
+/**
+ * Handle getting the source document or creating a new
+ * document with the correct mode and text.
+ *
+ */
+function showSourceText(editor, sourceText, selectedLocation) {
+  let doc = getDocument(selectedLocation.sourceId);
+  if (doc) {
+    editor.replaceDocument(doc);
+    return doc;
+  }
+
+  doc = editor.createDocument();
+  setDocument(selectedLocation.sourceId, doc);
+  editor.replaceDocument(doc);
+
+  setText(editor, sourceText.get("text"));
+  setMode(editor, sourceText);
+}
+
+/**
+ * Forces the breakpoint gutter to be the same size as the line
+ * numbers gutter. Editor CSS will absolutely position the gutter
+ * beneath the line numbers. This makes it easy to be flexible with
+ * how we overlay breakpoints.
+ */
+function resizeBreakpointGutter(editor) {
+  const gutters = editor.display.gutters;
+  const lineNumbers = gutters.querySelector(".CodeMirror-linenumbers");
+  const breakpoints = gutters.querySelector(".breakpoints");
+  breakpoints.style.width = lineNumbers.clientWidth + "px";
+}
+
+function clearDebugLine(editor, selectedFrame) {
+  if (selectedFrame) {
+    const line = selectedFrame.location.line;
+    editor.codeMirror.removeLineClass(line - 1, "line", "debug-line");
+  }
+}
+
+function setDebugLine(editor, selectedFrame, selectedLocation) {
+  if (selectedFrame && selectedLocation &&
+      selectedFrame.location.sourceId === selectedLocation.sourceId) {
+    const line = selectedFrame.location.line;
+    editor.codeMirror.addLineClass(line - 1, "line", "debug-line");
+  }
+}
+
+function editorHeight() {
```

Comment:
Cool, figured it wasn't entirely ready from your comment, but thought I'd give some review to help guide along. 😄